crates/faucet/build.rs

```diff
@@ -0,0 +1,92 @@
+use std::env;
+use std::fs::{self};
+use std::io::{self};
+use std::path::{Path, PathBuf};
+
+use miden_client::transaction::{TransactionKernel, TransactionScript};
+use miden_client::utils::Serializable;
+
+const ASSETS_DIR: &str = "assets";
+const ASM_DIR: &str = "asm";
+const ASM_TX_SCRIPTS_DIR: &str = "tx_scripts";
+
+/// Compile contents of asm directory into .txs files.
+fn main() {
+    // re-build when the MASM code changes
+    println!("cargo::rerun-if-changed={ASM_DIR}/");
+
+    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+    let source_dir = Path::new(&crate_dir).join(ASM_DIR);
+
+    let build_dir = env::var("OUT_DIR").unwrap();
+    let target_dir = Path::new(&build_dir).join(ASSETS_DIR);
+
+    compile_transaction_scripts(
+        &source_dir.join(ASM_TX_SCRIPTS_DIR),
+        &target_dir.join(ASM_TX_SCRIPTS_DIR),
+    );
+}
+
+/// Reads all MASM files from the `source_dir`, compiles each file individually into a TXS
+/// file, and stores the compiled files into the `target_dir`.
+///
+/// The source files are expected to contain executable programs.
+fn compile_transaction_scripts(source_dir: &Path, target_dir: &Path) {
+    fs::create_dir_all(target_dir).expect("should create target directory");
+    let assembler = TransactionKernel::assembler();
+
+    let masm_files = get_masm_files(source_dir).expect("should find MASM files");
+    for masm_file_path in masm_files {
+        // read the MASM file, parse it, and serialize the parsed AST to bytes
+        let code = assembler
+            .clone()
+            .assemble_program(masm_file_path.clone())
+            .expect("program should assemble correctly");
+        let script = TransactionScript::new(code);
+
+        let bytes = script.to_bytes();
+
+        let masm_file_name = masm_file_path.file_name().expect("file name should exist");
+        let mut txs_file_path = target_dir.join(masm_file_name);
+
+        // write the binary TXS to the output dir
+        txs_file_path.set_extension("txs");
+        fs::write(txs_file_path, bytes).expect("should write .txs file");
```

Comment:
nit: In miden-base we have similar logic, and there we suffix our files `.masb`. For reasons of consistency, I'd stick to one convention: https://github.com/0xMiden/miden-base/blob/643dd211eee1302d5a92ddf66b0402d9dcbc5c0e/crates/miden-lib/build.rs#L344

Reply:
I think one subtle difference here is that this is a direct serialization of `TransactionScript` struct, while in `miden-base` we serialize a `Program` struct which, after deserialization, gets converted into a `NoteScript` struct. So, we either should:

- Serialize `Program` with `.masb` extension, or
- Serialize `TransactionScript` with some other extension (i.e., `.txs`).

I'm fine with either approach tbh.

Reply:
Ah yes - I remember now, the first approach requires the user to know about the `Program` struct. So, I probably have a slight preference toward the `.txs` approach.

It may be good to have standard serialization defined for transaction and note scripts in `miden-base` (i.e., to/from file method) - but that's for another time.